contrib/swig/src/main/scala/edu/cmu/dynet/Expression.scala

```diff
@@ -0,0 +1,239 @@
+  *  * randomly (e.g. `randomNormal`)
+  *  * from [[edu.cmu.dynet.Model]] parameters (e.g. `parameter`)
+  *  * from other expressions (e.g. `softmax` and `pow`)
+  */
+object Expression {
+  import edu.cmu.dynet.internal.{dynet_swig => dn}
+
+  /** Private helper function for wrapping methods that get expressions from the computation
+    * graph */
+  private def makeExpr(f: internal.ComputationGraph => internal.Expression): Expression = {
+    val version = ComputationGraph.version
+    val expr = f(ComputationGraph.cg)
+    new Expression(expr)
+  }
+
+  def input(s: Float): Expression = makeExpr(cg => dn.input(ComputationGraph.cg, s))
+  def input(fp: FloatPointer): Expression = makeExpr(cg => dn.input(ComputationGraph.cg, fp.floatp))
+  def input(d: Dim, pdata: FloatVector): Expression =
+    makeExpr(cg => dn.input(cg, d.dim, pdata.vector))
+  def input(d: Dim, ids: UnsignedVector, data: FloatVector, defdata: Float = 0f) =
+    makeExpr(cg => dn.input(cg, d.dim, ids.vector, data.vector, defdata))
+
+  def parameter(p: Parameter): Expression = makeExpr(cg => dn.parameter(cg, p.parameter))
+  def constParameter(p: Parameter): Expression = makeExpr(cg => dn.const_parameter(cg, p.parameter))
+
+  def lookup(p: LookupParameter, index: Long) = makeExpr(cg => dn.lookup(cg, p.lookupParameter, index))
+  def lookup(p: LookupParameter, pindex: UnsignedPointer) =
+    makeExpr(cg => dn.lookup(cg, p.lookupParameter, pindex.uintp))
+  def constLookup(p: LookupParameter, index: Long) =
+    makeExpr(cg => dn.lookup(cg, p.lookupParameter, index))
+  // def constLookup
+  def lookup(p: LookupParameter, indices: UnsignedVector) =
+    makeExpr(cg => dn.lookup(cg, p.lookupParameter, indices.vector))
+
+
+  def zeroes(d: Dim) = makeExpr(cg => dn.zeroes(cg, d.dim))
+  def randomNormal(d: Dim) = makeExpr(cg => dn.random_normal(cg, d.dim))
+  def randomBernoulli(d: Dim, p: Float, scale: Float = 1.0f) = makeExpr(
+    cg => dn.random_bernoulli(cg, d.dim, p, scale))
+  def randomUniform(d: Dim, left: Float, right: Float) = makeExpr(
+    cg => dn.random_uniform(cg, d.dim, left, right))
+
+  /* ARITHMETIC OPERATIONS */
+
+  private type BinaryTransform = (internal.Expression, internal.Expression) => internal.Expression
+  private def binary(e1: Expression, e2: Expression, combiner: BinaryTransform) = {
+    e1.ensureFresh()
+    e2.ensureFresh()
+    val expr = combiner(e1.expr, e2.expr)
+    new Expression(expr)
+  }
+
+  private type UnaryTransform = internal.Expression => internal.Expression
+  private def unary(e: Expression, transformer: UnaryTransform) = {
+    e.ensureFresh()
+    new Expression(transformer(e.expr))
+  }
+
+  def exprMinus(e: Expression): Expression = unary(e, dn.exprMinus)
+  def exprPlus(e1: Expression, e2: Expression): Expression = binary(e1, e2, dn.exprPlus)
+  def exprPlus(e1: Expression, x: Float): Expression = unary(e1, e1 => dn.exprPlus(e1, x))
+  def exprPlus(x: Float, e2: Expression): Expression = exprPlus(e2, x)
```

Comment:
I was being cute and implementing it using the function above, but I changed it, because that's probably not a good idea